File: src/Haskell/X86Function.hs
{-# LANGUAGE MultiParamTypeClasses #-}
module X86Function where

import Backend
import X86Instr

data X86Function = X86Function { x86FunctionName:: String,
                                 x86FunctionBody:: [X86Instr] }

mkX86Function name body =
  X86Function { x86FunctionName = name,
                x86FunctionBody = body }

instance MachineFunction X86Function X86Instr where
  machineFunctionName = x86FunctionName
  machineFunctionBody = x86FunctionBody

  machineFunctionRename m sigma = error "needed later for register allocation"
  machineFunctionSpill m toSpill = error "needed later for register allocation"

File: src/Rust/x86/x86function.rs
use backend;
use ident::{Ident, Label};
use x86::X86Instr;

#[derive(Debug)]
pub struct X86Function {
    pub name: Label,
    pub body: Vec<X86Instr>,
    locals_in_stack: usize,
}

impl X86Function {
    pub fn new(name: Label, body: Vec<X86Instr>) -> X86Function {
        X86Function {
            name: name,
            body: body,
            locals_in_stack: 0,
        }
    }
}

impl backend::MachineFunction<X86Instr> for X86Function {
    fn name(&self) -> &Label {
        &self.name
    }

    fn body(&self) -> &Vec<X86Instr> {
        &self.body
    }

    fn spill(&mut self, _to_spill: &[Ident]) {
        panic!("needed later for register allocation")
    }

    fn rename(&mut self, _sigma: &Fn(Ident) -> Ident) {
        panic!("needed later for register allocation")
    }
}
